Expand Xnamespace

squishypinkelephant opened the pull request:
expand namespace capabilities to handle (all?) of the major extensions modify checks to enable functionality of a window manager running in the root namespace add in checks for superpower to function properly
try to make the allowMouseMotion permission consistent add globalkeyboard as a specific permission for keypress events sent to root namespace add randr and render permission settings

this *should* work for better isolation while also allowing everything to "just work" inside a WM

tested inside Xephyr using icewm. some programs ask for absurd things and will make minimizing resource access tricky
regardless, running a window manger in the root namespace, I can open firefix, gimp, krita, dolphin in independent namespaces. xkbcat and other still denied global keyboard by new keyword. 

it's a little crude and thrown together but it satisfies what i think is a perfect start for further improvements.

callmetango commented:
Just FYI because the discussion more or less augments your idea: [Ideas on XNamespace. · X11Libre · Discussion #440](https://github.com/orgs/X11Libre/discussions/440)

metux commented:
> some programs ask for absurd things and will make minimizing resource access tricky regardless, running a window manger in the root namespace, 

Window managers probably should always run in root NS, since they usually need a huge amount of control.
In the longer run, WMs should also become NS-aware, eg. showing which NS some client is running in, etc.

> xkbcat and other still denied global keyboard by new keyword.

This is one of the quite low level tools that probably belong into root NS. Same for various hotkey manager tools.

metux commented:
> Of course, no complaints here. I'd prefer things be more refined as well. In terms of total goals: getting "any" software running and functional in a non-root namespace while maintaining the original concepts: no access between namespaces, selective access to root resources. With as much as I've put in this far, I would say the goal is met 70%.

Well, 100% would be somewhat self-contradicting, since some clients are supposed to have control over others or global aspects of the Xserver, e.g window managers, input or output configuration tools, hotkey managers, ....
Perhaps we should instead talk about "user applications" - those clients that people use for some actual work (while putting the others into the "infrastructure" bin).

> As some already know, protecting the screen is all or nothing via mit-shm (and composite). 

Indeed. These things just haven't been fully finished yet.
Composite extension isn't supposed to be used by unprivileged clients (that would be a contradiction in itself), just by the privileged window manager / compositor (root NS).

> There's various random requests from screen capture tools to capture the mouse pointer and those requests will be either simply allowed (if they don't give away location hints like pulling only the pointer image) or restricted to mouse-motion. 

Those tools naturally belong into root / privileged NS, because it's their duty to touch privileged things.
A tricky question is how exactly do deal with browsers that wanna do screen sharing.

> As a stretch goal, Obs runs some checks through xcb instead of xlib and I'd like to get it working from inside a non-root namespace. I can get it most of the way by exposing shm and composite but the checks it runs don't trigger any alerts and the xcomposite extension gets disabled. 

What exactly does OBS do with xcomposite ?

> A nice to have would be a toggle and default quiet logs because Firefox at least loves to spam the absolute hell out of junk requests. And this is causing obvious lagging.

Yes, logging indeed needs improvements. For now it's pretty verbose in order to assist development as well as config/policy tuning.

metux commented:
> I've been working on moving the logic for adding/modifying/deleting namespaces into independent functions to make adding a protocol extension seamless.

Deleting namespaces is still an open topic - how shall we deal with existing clients ?
Just kill them off ? Refuse deletion of non-empty namespaces ?

squishypinkelephant committed:
expand namespace

squashed commit to clean up history

Signed-off-by: squishypinkelephant <squishypinkelephant@gmail.com>

squishypinkelephant committed:
client pid searching

fixes the annoyance with programs (firefox and others) spamming new namespaces
without breaking the new_ns function

Signed-off-by: squishypinkelephant <squishypinkelephant@gmail.com>

squishypinkelephant committed:
globally expose X_XIQueryDevice

Signed-off-by: squishypinkelephant <squishypinkelephant@gmail.com>

squishypinkelephant committed:
fix clientname complaints

Signed-off-by: squishypinkelephant <squishypinkelephant@gmail.com>

squishypinkelephant committed:
fix declaration complaint

Signed-off-by: squishypinkelephant <squishypinkelephant@gmail.com>

stefan11111 commented on Xext/namespace/config.c:
```diff
@@ -97,6 +121,8 @@ static void parseLine(char *line, struct Xnamespace **walk_ns)
         }
 
         curr = *walk_ns = select_ns(token);
+        // anything made from the config should be retained, this flag will serve dual purpose
```
Could you use C-style comments?

Same for the other c++-style comments.

stefan11111 commented on Xext/namespace/config.c:
```diff
@@ -25,9 +33,14 @@ struct Xnamespace ns_anon = {
+struct xorg_list client_list = { 0 };
+
 char *namespaceConfigFile = NULL;
+char *default_namespace;
```
Should this be `static const char*` instead?

squishypinkelephant committed:
minor cleanup

Signed-off-by: squishypinkelephant <squishypinkelephant@gmail.com>

squishypinkelephant committed:
Merge branch 'X11Libre:master' into master

squishypinkelephant committed:
minor additions and adjustments

move client name list to namespace struct
add client path list as alternative to just name. stricter control.
add a few atoms to resource check
remove the namespace list print function. only needed it for debugging, no reason to leave it
fix a redundant printf
add in tab delim to match 2683973

Signed-off-by: squishypinkelephant <squishypinkelephant@gmail.com>

squishypinkelephant commented:
taking the idea from https://github.com/orgs/X11Libre/discussions/63 
using paths for strict assignment works but i can't imagine it working out for nix/guix
